### PR TITLE
[DDW-368]: Fetch slot and epoch info from the API

### PR DESCRIPTION
**Pull request by yakovkaravelov**
This PR implemented slot and epoch length variables configuration fetching API instead of constants.
[Jira Ticket](https://jira.iohk.io/browse/DDW-368)

## Todos
- [x] Fetch slot and epoch length info from the API instead of constants

## Testing Checklist
- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1598319317026400)

### App
- [x] Do regression testing

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Review by mioriohk — APPROVED** [review]
Looks good 👍

**Review by ManusMcCole — APPROVED** [review]

**Comment by nikolaglumac** [comment]
@daniloprates please review this one today 🙏

File: source/renderer/app/api/api.js [hunk 1]
```diff
@@ -1657,11 +1655,7 @@ export default class AdaApi {
           ? {
               // N+2 epoch
               epochNumber: nextEpochNumber ? nextEpochNumber + 1 : null,
-              epochStart: nextEpochStartTime
-                ? moment(nextEpochStartTime)
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:1661]: @yakovkaravelov why did you remove this? This is start time of the N+2 epoch and not the N+1 (next) one?

File: source/renderer/app/api/utils/patchAdaApi.js [hunk 2]
```diff
@@ -73,11 +70,7 @@ export default (api: AdaApi) => {
         futureEpoch: {
           // N+2 epoch
           epochNumber: nextEpochNumber ? nextEpochNumber + 1 : null,
-          epochStart: nextEpochStartTime
-            ? moment(nextEpochStartTime)
```
Comment (nikolaglumac) [inline source/renderer/app/api/utils/patchAdaApi.js:77]: @yakovkaravelov why did you remove this? This is start time of the N+2 epoch and not the N+1 (next) one?

File: source/renderer/app/config/epochsConfig.js [hunk 3]
```diff
@@ -8,29 +8,27 @@ export const START_TIME_STAGING = 1506450213;
...
-export const EPOCH_LENGTH_ITN = SLOT_DURATION_ITN * TOTAL_SLOTS_ITN; // 1 day / 5 minutes (isDev || isTest || isIncentivizedTestnetSelfnode = true) | unit: seconds
+export const DEFAULT_EPOCH_COUNTDOWN_INTERVAL = 1 * 1000; // 1 second | unit: milliseconds;
 
-export const EPOCH_COUNTDOWN_INTERVAL = 1 * 1000; // 1 second | unit: milliseconds;
-
-export const getTotalSlots = () => {
-  if (isIncentivizedTestnet) return TOTAL_SLOTS_ITN;
-  if (isShelleyTestnet) return TOTAL_SLOTS_STN;
-  return TOTAL_SLOTS;
+export const getDefaultSlotLength = () => {
```
Comment (nikolaglumac) [inline source/renderer/app/config/epochsConfig.js:24]: @yakovkaravelov can't we eliminate the need for these hardcoded values? This is not safe... We should only use the ones we get from the Api.

File: source/renderer/app/config/epochsConfig.js [hunk 4]
```diff
@@ -8,29 +8,27 @@ export const START_TIME_STAGING = 1506450213;
...
+  if (isIncentivizedTestnet) return DEFAULT_SLOT_LENGTH_ITN;
+  if (isShelleyTestnet) return DEFAULT_SLOT_LENGTH_STN;
+  return DEFAULT_SLOT_LENGTH;
 };
 
-export const getEpochLength = () => {
-  if (isIncentivizedTestnet) return EPOCH_LENGTH_ITN;
-  if (isShelleyTestnet) return EPOCH_LENGTH_STN;
-  return EPOCH_LENGTH;
+export const getDefaultEpochLength = () => {
```
Comment (nikolaglumac) [inline source/renderer/app/config/epochsConfig.js:30]: @yakovkaravelov can't we eliminate the need for these hardcoded values? This is not safe... We should only use the ones we get from the Api.

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@yakovkaravelov please check my comments.

File: source/renderer/app/api/api.js (hunk 1 above)
Comment (yakovkaravelov) [reply source/renderer/app/api/api.js:1661]: @nikolaglumac I didn't remove this.
This is calculated in another file.

File: source/renderer/app/api/utils/patchAdaApi.js (hunk 2 above)
Comment (yakovkaravelov) [reply source/renderer/app/api/utils/patchAdaApi.js:77]: @nikolaglumac I didn't remove this.
This is calculated in another file.

File: source/renderer/app/config/epochsConfig.js (hunk 3 above)
Comment (yakovkaravelov) [reply source/renderer/app/config/epochsConfig.js:24]: @nikolaglumac This is just initialization part before api returns response.
As soon as API returns values we use them instead of these values.

File: source/renderer/app/config/epochsConfig.js (hunk 4 above)
Comment (yakovkaravelov) [reply source/renderer/app/config/epochsConfig.js:30]: @nikolaglumac This is just initialization part before api returns response.
As soon as API returns values we use them instead of these values.

**Comment by yakovkaravelov** [comment]
@nikolaglumac Please review my replies.

File: source/renderer/app/api/api.js (hunk 1 above)
Comment (nikolaglumac) [reply source/renderer/app/api/api.js:1661]: Not as far as I can see: https://github.com/input-output-hk/daedalus/pull/2165/files#diff-6aa9e91ae277a5d21b8e1bb0b03ffafbR1633

File: source/renderer/app/api/utils/patchAdaApi.js (hunk 2 above)
Comment (nikolaglumac) [reply source/renderer/app/api/utils/patchAdaApi.js:77]: Not as far as I can see: https://github.com/input-output-hk/daedalus/pull/2165/files#diff-6aa9e91ae277a5d21b8e1bb0b03ffafbR1633

File: source/renderer/app/config/epochsConfig.js (hunk 3 above)
Comment (nikolaglumac) [reply source/renderer/app/config/epochsConfig.js:24]: Yes but this could result in false data - e.g. if epoch length changes - it can change dynamically.
I would prefer to remove any default data and have Daedalus not show parts of the screen where this data is used until it is loaded from the Api!

File: source/renderer/app/config/epochsConfig.js (hunk 4 above)
Comment (nikolaglumac) [reply source/renderer/app/config/epochsConfig.js:30]: Yes but this could result in false data - e.g. if epoch length changes - it can change dynamically.
I would prefer to remove any default data and have Daedalus not show parts of the screen where this data is used until it is loaded from the Api!

**Comment by nikolaglumac** [comment]
@yakovkaravelov please review my replies.

File: source/renderer/app/api/api.js (hunk 1 above)
Comment (yakovkaravelov) [reply source/renderer/app/api/api.js:1661]: @nikolaglumac https://github.com/input-output-hk/daedalus/pull/2165/files#diff-faefd5629e4c983973ee387b12021c59R483
It is calculated here because epochLength should be observable variable of networkStatusStore.

File: source/renderer/app/api/utils/patchAdaApi.js (hunk 2 above)
Comment (yakovkaravelov) [reply source/renderer/app/api/utils/patchAdaApi.js:77]: @nikolaglumac Same as above comment.

File: source/renderer/app/config/epochsConfig.js (hunk 3 above)
Comment (yakovkaravelov) [reply source/renderer/app/config/epochsConfig.js:24]: @nikolaglumac Fixed this.

File: source/renderer/app/config/epochsConfig.js (hunk 4 above)
Comment (yakovkaravelov) [reply source/renderer/app/config/epochsConfig.js:30]: @nikolaglumac Fixed this. Thanks.

File: source/renderer/app/api/api.js (hunk 1 above)
Comment (nikolaglumac) [reply source/renderer/app/api/api.js:1662]: @yakovkaravelov I do not agree - please see attached screenshot.
The "next" and the "future" epochs can NOT have the same epoch start time...
Future epoch starts after the next epoch ends. Which means it's start time is the start time of the next epoch PLUS a length of a single epoch.

see

File: source/renderer/app/api/utils/patchAdaApi.js (hunk 2 above)
Comment (nikolaglumac) [reply source/renderer/app/api/utils/patchAdaApi.js:77]: @yakovkaravelov https://github.com/input-output-hk/daedalus/pull/2165/files#r484325653

**Comment by yakovkaravelov** [comment]
@nikolaglumac Fixed things.

File: source/renderer/app/api/network/types.js [hunk 5]
```diff
@@ -54,7 +54,7 @@ export type GetNetworkInfoResponse = {
   localTip: TipInfo,
   networkTip: ?TipInfo,
   nextEpoch: ?NextEpoch,
-  futureEpoch: ?FutureEpoch,
+  futureEpoch?: ?FutureEpoch,
```
Comment (nikolaglumac) [inline source/renderer/app/api/network/types.js:57]: @yakovkaravelov shouldn't we remove `futureEpoch` from this type declaration?

Comment (nikolaglumac) [reply source/renderer/app/api/network/types.js]: @yakovkaravelov fixed in https://github.com/input-output-hk/daedalus/pull/2165/commits/5f8b08992a9e2ccc49eb834beb46b2149b16c8ce

**Review by nikolaglumac — APPROVED** [review]
Great work @yakovkaravelov 💯
